packages/mui-joy/src/Button/Button.tsx

```diff
@@ -1,21 +1,248 @@
 import * as React from 'react';
 import PropTypes from 'prop-types';
+import clsx from 'clsx';
+import { unstable_capitalize as capitalize, unstable_useForkRef as useForkRef } from '@mui/utils';
+import { useButton } from '@mui/base/ButtonUnstyled';
+import composeClasses from '@mui/base/composeClasses';
+import { styled, useThemeProps } from '../styles';
+import { ExtendButton, ButtonTypeMap, ButtonProps } from './ButtonProps';
+import buttonClasses, { getButtonUtilityClass } from './buttonClasses';
 
-export type ButtonProps = JSX.IntrinsicElements['button'];
+const useUtilityClasses = (ownerState: ButtonProps & { focusVisible: boolean }) => {
+  const { color, disabled, focusVisible, focusVisibleClassName, fullWidth, size, variant } =
+    ownerState;
 
-const Button = (props: ButtonProps) => {
-  return <button type="button" {...props} />;
+  const slots = {
+    root: [
+      'root',
+      disabled && 'disabled',
+      focusVisible && 'focusVisible',
+      fullWidth && 'fullWidth',
+      `variant${capitalize(variant!)}`,
+      `color${capitalize(color!)}`,
+      `size${capitalize(size!)}`,
+    ],
+  };
+
+  const composedClasses = composeClasses(slots, getButtonUtilityClass, {});
+
+  if (focusVisible && focusVisibleClassName) {
+    composedClasses.root += ` ${focusVisibleClassName}`;
+  }
+
+  return composedClasses;
 };
 
+const ButtonRoot = styled('button', {
+  name: 'MuiButton',
+  slot: 'Root',
+  overridesResolver: (props, styles) => {
+    const { ownerState } = props;
+
+    return [
+      styles.root,
+      styles[`variant${capitalize(ownerState.variant)}`],
+      styles[`color${capitalize(ownerState.color)}`],
+      styles[`size${capitalize(ownerState.size)}`],
+      ownerState.fullWidth && styles.fullWidth,
+    ];
+  },
+})<{ ownerState: ButtonProps }>(({ theme, ownerState }) => {
+  const colorPalette = theme.vars.palette[ownerState.color || 'brand'];
```

Comment:
From a DX perspective, I wonder if we shouldn't reverse the tradeoff. It seems that the right path (using the CSS variable) is more painful than accessing the raw value.

```suggestion
  const colorPalette = theme.palette[ownerState.color || 'brand'];
```

Reply:
This is what I propose in https://github.com/mui-org/material-ui/issues/27651 under "Decision on the API".

2 things about using `vars`
- I consider CSS variables as a feature on top of ThemeProvider, so adding `vars` instead of replace the theme makes more sense.
- When we have to update `mui-material` to support CSS variables, we can incrementally change the code in each component and it should not break other components.

Anyway, I will try your method and share the finding again. If it is better, we can change Joy to use it.